### PR TITLE
fix: variable not found #26144

### PR DESCRIPTION
**Pull request by sqewad**
1. variable not found when accessing a nested attr with empty value

> [!IMPORTANT]
>
> 1. Make sure you have read our [contribution guidelines](https://github.com/langgenius/dify/blob/main/CONTRIBUTING.md)
> 1. Ensure there is an associated issue and you have been assigned to it
> 1. Use the correct syntax to link this PR: `Fixes #<issue number>`.

## Summary

fix: #26144
Distinguish between the case where a variable exists but its value is empty and the case where the variable does not exist.

## Screenshots

| Before | After |
|--------|-------|
| Weixin Image_20250924084836_137 | Weixin Image_20250924084833_138 |

## Checklist

- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [x] I understand that this PR may be closed in case there was no previous discussion or issues. (This doesn't apply to typos!)
- [x] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [x] I've updated the documentation accordingly.
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

### CODE REVIEW

File: api/core/workflow/entities/variable_pool.py [hunk 1]
```diff
@@ -187,8 +187,8 @@ def _extract_value(self, obj: Any):
     def _get_nested_attribute(self, obj: Mapping[str, Any], attr: str):
         """Get a nested attribute from a dictionary-like object."""
         if not isinstance(obj, dict):
-            return None
-        return obj.get(attr)
+            return False, None
+        return attr in obj, obj.get(attr)
```
[the remark this answers is not in the document]
Comment (sqewad) [reply api/core/workflow/entities/variable_pool.py]: i think using "in" is more readable. there aren’t many performance requirements here, right?

File: api/core/workflow/entities/variable_pool.py [hunk 2]
```diff
@@ -187,8 +187,8 @@ def _extract_value(self, obj: Any):
     def _get_nested_attribute(self, obj: Mapping[str, Any], attr: str):
```
Comment (QuantumGhost) [inline api/core/workflow/entities/variable_pool.py]: A minor suggestion: instead of returning `tuple[bool, Any]`, consider changing the signature of this function to `tuple[Any, bool]`, and adding a short description about the second value in the tuple to the docstring.

**Review by QuantumGhost — COMMENTED** [review]
Hi @sqewad, thank you for the contribution!

The overall changes look good to me. However, I'd suggest adding unit tests for both the `get` and `_get_nested_attribute` methods to cover the specific edge case you fixed. 
This will help prevent regression and ensure the nested attribute handling works correctly in various scenarios.

This would make the fix more robust and maintainable.

File: api/core/workflow/entities/variable_pool.py (hunk 2 above)
Comment (sqewad) [reply api/core/workflow/entities/variable_pool.py]: sure I'll do it, thanks for the suggestion

File: api/core/workflow/entities/variable_pool.py [hunk 3]
```diff
@@ -184,11 +184,22 @@ def _extract_value(self, obj: Any):
...
+        """
+        Get a nested attribute from a dictionary-like object.
+
+        Args:
+            obj: The dictionary-like object to search.
+            attr: The key to look up.
+
+        Returns:
+            tuple:
+                - A boolean indicating whether the attribute exists in the object.
```
Comment (QuantumGhost) [inline api/core/workflow/entities/variable_pool.py]: It seems that the order of elements in returned tuple is not updated.

IMO it's better to return the two values as 
`None, False` and `obj.get(attr), attr in obj`

Comment (sqewad) [reply api/core/workflow/entities/variable_pool.py]: Thanks for pointing it out, I’ve updated the tuple order as suggested.

File: api/core/workflow/entities/variable_pool.py [hunk 4]
```diff
@@ -184,11 +184,22 @@ def _extract_value(self, obj: Any):
...
+
+        Returns:
+            tuple:
+                - The value of the attribute if found, otherwise None.
+                - A boolean indicating whether the attribute exists in the object.
+        """
         if not isinstance(obj, dict):
-            return None
-        return obj.get(attr)
+            return None, False
```
Comment (laipz8200) [inline api/core/workflow/entities/variable_pool.py]: return a tuple is hard to read, please consider returning a NoneSegment if the key exists but value is None.

you can use `variable_factory.build_segment(None)` to get one.

**Review by laipz8200 — CHANGES_REQUESTED** [review]

File: api/core/workflow/entities/variable_pool.py (hunk 4 above)
Comment (sqewad) [reply api/core/workflow/entities/variable_pool.py]: good idea, woking on it

**Review by laipz8200 — APPROVED** [review]
